### DISCUSSION TITLE
BulkInsert limitations bulk size

### DISCUSSION THREAD

**Discussion by dekke046** [sijms/go-ora#291] [Ideas]
Hi,

We come from a lot of shell scripting (bash) where we use Oracle SQL*Loader to import data into various tables.
SQL*Loader is fast, extremely fast, but the eco system of using shell scripting has had its time.

Using go-ora is perfect to build small and leightweight microservices (docker) to do many activities, but when it comes to bulkInserts it is limited. On production LAN I cant create bigger array's than 200 - 250, with bigger arrays I will have EOF or broken pipe errors.

Doing the same on a localhost I can go to 40000 as an array size without errors, so defitely the network is a factor (latency/speed) which has influence on the max BulkInsert sizes, but that does not mean the network is unstable.

It would be really helpful if BulkInsert can handle larger array sizes. inserting 2.2 million records with SQL*Loader takes roughly 6 seconds. Using go-ora with BulkInsert and an arraysize of 200 takes more that 3 minutes. Which practically means we cant use it.

For reference:
https://github.com/sijms/go-ora/issues/248

I am really looking forward to some ideas to either use the existing go-ora for large inserts the same way as SQL*Loader does, or introduce some new functions.

Thanks,

Maarten

**Comment by sijms** [comment]
you make bulk insert for 200 rows in 3 minutes !
[this disscussion](https://github.com/sijms/go-ora/discussions/163#discussioncomment-2233190) explain bulk insert of around 8000 rows in 2 seconds

**Reply by dekke046** [reply]
Hi @sijms,

Thanks for your reply!

What I currently do is processing records from a file (2.2 million rows) and import them into the database with BulkInsert.
The variable **maxRecordsInsert** is set to 200.

So the BulkInsert will inserter 200 rows and will do that in a reasonable time:
2023/01/13 15:15:42 200 rows inserted: 10.908349ms
2023/01/13 15:15:42 200 rows inserted: 15.079309ms
2023/01/13 15:15:42 200 rows inserted: 9.369048ms
2023/01/13 15:15:42 200 rows inserted: 9.643844ms
2023/01/13 15:15:42 200 rows inserted: 8.384206ms

So the insert of small sizes (maxRecordsInsert) goes fine on our production LAN.
If I increase 200 to a higher number, lets say 1000 it does not even come to an insert, immediately it gives an EOF error.
There are more than 2.2 million rows in the file.

Lets have a look at the code snippet below:
```
// Slices for bulkinsert
colA := []driver.Value{}
colB := []driver.Value{}
........
	for sc.Scan() {
		values := strings.Split(sc.Text(), "|")
		colA = append(colA, values[0])
		colB = append(colB, values[1])
		lines++
		if lines >= maxRecordsInsert {
			t := time.Now()
			result, err := conn.BulkInsert(insertStatement, len(colA), colA, colB)
			if err != nil {
				return err
			}
			rowsAffected, _ := result.RowsAffected()

			log.Printf("%d rows inserted: %v\n", rowsAffected, time.Now().Sub(t))
			lines = 0
                        // Start over with empty slices
			colA = nil
			colB = nil
		}
	}
........
```
If I do this process the old way, so a bash script with Oracle SQL*Loader and a CTL file on the same production LAN I can load all 2.2 million records into the database in less than 7 seconds.

As a workaround solution I now use go_ora with BulkInsert to insert the records in batches of 200 into a temporary table, when done I truncate the real table and copy from temporry table to the real production table.

Maarten

**Comment by sijms** [comment]
from the example bulk insert I comment insert data and delete data functions, change the rowNum to **100000**
this is the result
```golang
Finish create table GOORA_TEMP_VISIT : 39.953437ms
100000 rows inserted: 3.563793612s
Finish drop table:  519.747375ms
```

if I change rowNum to **1000000**
```golang
Finish create table GOORA_TEMP_VISIT : 41.49456ms
1000000 rows inserted: 29.867525222s
Finish drop table:  3.984865642s
```

**environment:**
server: 12.2c docker image
client on the same device (macOS)
connection is basic without encryption or data integrity using TCP

**Reply by dekke046** [reply]
@sijms.

You are right that is the same behaviour I am experiencing as well.
Running it all local I can go up to very high numbers, since there is no real network involved at all.

The same numbers are not possible with a real network, in our case a production datacenter LAN.
So we can conlude that as soon as there are some form of network latencies or delays BulkInsert will not work with higher numbers of rows.
Using SQL*Loader on the same network is not causing such issues.
So there must be something on the low level (TCP, buffers, timeouts) of the driver that causes the behaviour.
Are there specific settings being used on TCP level of the go-ora driver?

I will also see if I can capture anything with TCPDump

**Reply by cjbj** [reply]
Some references: 
- https://docs.oracle.com/en/database/oracle/oracle-database/21/netag/optimizing-performance.html
- https://www.oracle.com/docs/tech/database/oow19-net-con4641.pdf

**Comment by sijms** [comment]
note that I get error If I use encryption service with row number > 700 rows
and this is because bug (problem) in encryption not bulk insert

**Reply by sijms** [reply]
The Bug related to Data Packet Encryption is fixed in v2.5.24

**Comment by sijms** [comment]
I think this will help [BulkCopy](https://docs.oracle.com/database/121/ODPNT/featBulkCopy.htm#ODPNT212)

**Comment by sijms** [comment]
I test BulkCopy using Direct Path still time is same as Bulk Insert, sometimes more
